File: src/providers/openai-realtime/rules/avoid-dated-preview-snapshots.ts
import {
  collectOpenAIRealtimeUrlVarNames,
  collectStringVarValues,
  isOpenAIRealtimeNewWebSocket,
  resolveStringValue,
} from '../utils.js';

const DATED_PREVIEW_MODEL_PATTERN = /model=[^&'"`]*-preview-\d{4}-\d{2}-\d{2}/;

const rule = {
  meta: {
    type: 'suggestion',
    docs: {
      description: 'OpenAI Realtime connections should not pin to a dated preview model snapshot',
      category: 'correctness',
      docsUrl: 'https://developers.openai.com/api/docs/api-reference/realtime-sessions',
      rationale:
        "Dated preview snapshots are the category of model id OpenAI has historically retired with advance-notice sunset windows. Pinning to a dated preview snapshot rather than the floating GA alias (or a current dated GA snapshot) maximizes exposure to an eventual retirement, with no code path to detect or gracefully react to a model_not_found-style error if/when that happens.",
      recommended: true,
    },
    messages: {
      datedPreviewSnapshot:
        'This Realtime connection is pinned to a dated preview model snapshot instead of the GA alias.',
    },
    schema: [],
  },
  create(context: any) {
    let stringVarValues = new Map<string, string>();
    let urlVarNames = new Set<string>();

    return {
      Program(node: any) {
        stringVarValues = collectStringVarValues(node);
        urlVarNames = collectOpenAIRealtimeUrlVarNames(node);
      },

      NewExpression(node: any) {
        if (!isOpenAIRealtimeNewWebSocket(node, urlVarNames)) return;

        const urlArg = node.arguments?.[0];
        const urlString = resolveStringValue(urlArg, stringVarValues);
        if (!urlString) return;

        if (DATED_PREVIEW_MODEL_PATTERN.test(urlString)) {
          context.report({ node: urlArg, messageId: 'datedPreviewSnapshot' });
        }
      },
    };
  },
};

export const openaiRealtimeAvoidDatedPreviewSnapshotsRule = rule;

File: src/providers/openai-realtime/rules/python/avoid_dated_preview_snapshots.py
"""openai-realtime-avoid-dated-preview-snapshots

Parity with the JS rule: a Realtime connection pinned to a dated preview
model snapshot (rather than the floating GA alias) maximizes exposure to an
eventual retirement with no code path to detect it.
"""
from __future__ import annotations

import ast
import re
import sys
from pathlib import Path

from ast_utils import loc

_PROVIDER = Path(__file__).resolve().parents[2]
if str(_PROVIDER) not in sys.path:
    sys.path.insert(0, str(_PROVIDER))
from utils import (  # noqa: E402
    collect_realtime_url_var_names,
    collect_string_var_values,
    is_realtime_connect_call,
    resolve_string_value,
)

RULE_KEY = "openai-realtime-avoid-dated-preview-snapshots"

DATED_PREVIEW_MODEL_RE = re.compile(r"model=[^&'\"`]*-preview-\d{4}-\d{2}-\d{2}")


def check(tree: ast.AST, path: str, source: str) -> list[dict]:
    string_vars = collect_string_var_values(tree)
    url_var_names = collect_realtime_url_var_names(tree)
    out: list[dict] = []

    for node in ast.walk(tree):
        if not isinstance(node, ast.Call) or not is_realtime_connect_call(node, url_var_names):
            continue
        url_arg = node.args[0]
        url_string = resolve_string_value(url_arg, string_vars)
        if not url_string:
            continue
        if DATED_PREVIEW_MODEL_RE.search(url_string):
            line, col, _, _ = loc(url_arg)
            out.append({"file": path, "line": line, "column": col, "ruleKey": RULE_KEY})

    return out
